### PR TITLE
Add retry logic for Azure service management operations

### LINKED ISSUE #24: Intermittent Build Failure
Hi Vlad,

Sometimes when I run packer, I hit a failure due to "Windows Azure is currently performing an operation on this deployment that requires exclusive access".  I'm sorry I don't have more logging information.  I retried running packer and the image was successfully created.  Is there retry logic around "Deleting Temporary Directory"?

thanks

mary

==> azure: Executing OS generalization...
==> azure: Stopping Temporary Azure VM...
==> azure: Creating Azure Image. If Succeed This Will Remove the Temorary VM...
==> azure: Error Creating Azure Image: Manager.ExecuteSync: Windows Azure is currently performing an operation on this deployment that requires exclusive access.
==> azure: Removing Temporary Azure Service and It's Deployments If Any...
==> azure: Deleting Temporary Certificate...
    azure: Deleting Temporary Directory...
Build 'azure' errored: Error Creating Azure Image: Manager.ExecuteSync: Windows Azure is currently performing an operation on this deployment that requires exclusive access.

==> Some builds didn't complete successfully and had errors:
--> azure: Error Creating Azure Image: Manager.ExecuteSync: Windows Azure is currently performing an operation on this deployment that requires exclusive access.

==> Builds finished but no artifacts were created.
Packer has finished

ERROR: Packer build was unsuccessful

### LINKED ISSUE #25: Azure Packer Build Request timed out
azure output will be in this color.

==> azure: Preparing builder...
    azure: Getting subscr info...
    azure: Creating a New Driver...
    azure: Creating a New Request Manager...
==> azure: Validating Azure Options...
    azure: Checking Storage Account...
Build 'azure' errored: Some Azure options failed: Get https://management.core.windows.net/<subs_id>/services/storageservices/operations/isavailable/us1preprodall: dial tcp 157.55.195.251:443: connection timed out

==> Some builds didn't complete successfully and had errors:
--> azure: Some Azure options failed: Get https://management.core.windows.net/04dc0d13-8d80-4781-b4b9-603708ce871c/services/storageservices/operations/isavailable/us1preprodall: dial tcp 157.55.195.251:443: connection timed out

==> Builds finished but no artifacts were created.

Im behind an enterprise proxy. I have set up env variables http_proxy and https_proxy.

Is azure packer using the proxy settings? If not, how can I make it use the proxy settings?

Regards

### PR DESCRIPTION
**Pull request by paulmey**
Removed separate 'driver' concept, collapsed it into the 'manager' object.
Changed from io.Reader to []byte for carrying the request payloads.
Use same api version for all requests.

Fixes #24 
Fixes #25

cc: @ahmetalpbalkan @Snesha @jeffmendoza

### CODE REVIEW

File: packer/builder/azure/driver_restapi/request/RetryPolicy.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+			return (err.Code == "BadRequest" && strings.Contains(err.Message, "is currently in use by")) ||
+				(err.Code == "ConflictError" && strings.Contains(err.Message, "that requires exclusive access"))
+		}, 10*time.Second, 100),
+	}
+}
+
+type retryRule func(model.AzureError) (bool, time.Duration)
+type retryPolicy []retryRule
+
+func (rules *retryPolicy) ShouldRetry(err model.AzureError) (shouldRetry bool, backoff time.Duration) {
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: * is probably redundant

File: packer/builder/azure/driver_restapi/request/RetryPolicy.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+)
+
+func defaultRetryPolicy() retryPolicy {
+	return retryPolicy{
+		exponentialBackoff("Throttling", func(err model.AzureError) bool {
+			return err.Code == "TooManyRequests"
+		}, 5*time.Second, 2*time.Minute, 0),
+		constantBackoff("MissingOrIncorrectVersionHeader", func(err model.AzureError) bool {
+			return err.Code == "MissingOrIncorrectVersionHeader" // shouldn't occur, since we always add it
+		}, 0, 100),
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: 100 seems a bit high, especially if you don't stop. this error shouldn't be retriable at all,.

File: packer/builder/azure/driver_restapi/request/RetryPolicy.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	for _, rule := range []retryRule(*rules) {
+		shouldRetry, backoff = rule(err)
+		if shouldRetry {
+			return
+		}
+	}
+	return false, 0
+}
+
+func constantBackoff(name string, match func(model.AzureError) bool, backoff time.Duration, maxRetries int) retryRule {
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: `type matchRule func(model.AzureError) bool ` could be useful as it also repeats a lot

**Comment by ahmetalpbalkan** [comment]
realized we're actually maintaining 2 azure management SDKs in Go. that probably shouldn't be the case. Consider improving the official one we have.

**Comment by paulmey** [comment]
yep, working on that too

File: packer/builder/azure/driver_restapi/request/RetryPolicy.go (hunk 2 above)
Comment (paulmey) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: Yeah, Azure shouldn't be throwing it either... The historic comment was that Azure sometimes gives you this error, apparently by mistake. Do I have your vote for removing this thing?

File: packer/builder/azure/driver_restapi/request/RetryPolicy.go (hunk 3 above)
Comment (paulmey) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: :+1:

File: packer/builder/azure/driver_restapi/request/RetryPolicy.go (hunk 1 above)
Comment (paulmey) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: which * exactly?

Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: there's only one `*` on line 37. :smile:

File: packer/builder/azure/driver_restapi/request/RetryPolicy.go (hunk 2 above)
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: > Azure sometimes gives you this error, apparently by mistake.

loool

Comment (paulmey) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: yep... an erroneous error... I don't think I've ever seen it. I would like to remove it here and file bugs for RDFE if anyone runs into it?

Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: Makes sense.

File: packer/builder/azure/driver_restapi/request/RetryPolicy.go (hunk 1 above)
Comment (paulmey) [inline packer/builder/azure/driver_restapi/request/RetryPolicy.go]: :ok_hand:

**Comment by jeffmendoza** [comment]
:+1:

**Pull request merged**